src/python/pants/backend/java/compile/javac.py

```diff
@@ -118,7 +126,23 @@ async def compile_java_source(
             exit_code=0,
         )
 
-    lockfile = await Get(CoursierResolvedLockfile, CoursierResolveKey, request.resolve)
+    coordinates = Coordinates(
+        (
+            Coordinate(
+                group=(dep[JvmArtifactGroupField].value or ""),
+                artifact=(dep[JvmArtifactArtifactField].value or ""),
+                version=(dep[JvmArtifactVersionField].value or ""),
+            )
+            for item in CoarsenedTargets(request.component.dependencies).closure()
+            for dep in item.members
+            if dep.has_fields((JvmArtifactGroupField, JvmArtifactArtifactField))
+        )
+    )
+
+    unfiltered_lockfile = await Get(CoursierResolvedLockfile, CoursierResolveKey, request.resolve)
+    lockfile = await Get(
+        CoursierResolvedLockfile, FilterDependenciesRequest(coordinates, unfiltered_lockfile)
+    )
```

Comment:
This change should be applied to `scalac` as well, so probably worth extracting something reusable for the `JvmArtifact` -> `Coordinate` conversion... and possibly making that change here as well.

Reply:
@stuhood see 0b1db21 and e47e5d0.